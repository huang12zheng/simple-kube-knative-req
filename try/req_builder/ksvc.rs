#[macro_use]
extern crate educe;

use kube::core::DynamicObject;
use serde_json::{json, Value};

#[derive(Educe)]
#[educe(Default, Debug)]
struct KnativeSpec {
    #[educe(Default = "serving.knative.dev")]
    group: String,
    version: String,
    #[educe(Default = "Service")]
    kind: String,
    // #[educe(Defaul(ignore))]
    #[educe(Default = "my-coder")]
    container_name: String,
    #[educe(Default = "hzgood/my-coder")]
    svc_image: String,
    #[educe(Default = 8080)]
    port: i32,
    env: Value,
}

trait IntoDynamicObject {
    fn gv(&self) -> String;
    fn into_do(&self) -> DynamicObject;
}
// ! impl KnativeSpec {
// !     /// Returns the gv of this [`KnativeSpec`].
// !     fn gv(&self) -> String {
// !         if self.version == "" {
// !             self.group.clone()
// !         } else {
// !             format!("{}/{}", self.group, self.version)
// !         }
// !     }
// !     pub fn ksvc(self) -> anyhow::Result<DynamicObject> {
// !         let ksvc: DynamicObject = serde_json::from_value(json!({
// !         "apiVersion": self.gv(),
// !         "kind": self.kind,
// !         "metadata": {
// !             // "name": format!("{}",container_name),
// !             "name": self.container_name,
// !         },
// !         "spec": {
// !             "template":{
// !                 "spec": {
// !                     "containers": [{
// !                         "ports": [
// !                             {
// !                                 "containerPort": self.port
// !                             }
// !                         ],
// !                         "image": self.svc_image,
// !                         "env": self.env
// !                     }],
// !                 }
// !             }
// !         }
// !         }))?;
// !         Ok(ksvc)
// !     }
// ! }
impl IntoDynamicObject for KnativeSpec {
    /// Returns the gv of this [`KnativeSpec`].
    fn gv(&self) -> String {
        if self.version == "" {
            self.group.clone()
        } else {
            format!("{}/{}", self.group, self.version)
        }
    }
    fn into_do(&self) -> DynamicObject {
        let ksvc: DynamicObject = serde_json::from_value(json!({
        "apiVersion": self.gv(),
        "kind": self.kind,
        "metadata": {
            // "name": format!("{}",container_name),
            "name": self.container_name,
        },
        "spec": {
            "template":{
                "spec": {
                    "containers": [{
                        "ports": [
                            {
                                "containerPort": self.port
                            }
                        ],
                        "image": self.svc_image,
                        "env": self.env
                    }],
                }
            }

        }
        }))
        .unwrap();
        // Ok(ksvc)
        ksvc
    }
}

#[test]
fn spec_works() {
    let spec = KnativeSpec {
        ..Default::default()
    };
    dbg!(spec);
    // assert_eq!(2 + 2, 4);
}

#[test]
fn show_into_do() {
    let spec = KnativeSpec {
        container_name: "my-coder".to_owned(),
        ..Default::default()
    };
    let ksvc = spec.into_do();
    insta::assert_debug_snapshot!(ksvc)
}
fn main() {}
