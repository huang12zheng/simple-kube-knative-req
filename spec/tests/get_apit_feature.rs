//! run test use `INSTA_UPDATE=1 cargo test`
use std::convert::Infallible;

use async_trait::async_trait;
use cucumber::{gherkin::Step, given, then, World, WorldInit};
use envconfig::Envconfig;
use insta::assert_debug_snapshot;
use kube_do_spec::ksvc::ksvc::KnativeSpec;
use kube_do_spec::ksvc::ksvc_api::{KnativeSpecIntoApi, KnativeSpecWrapper};

#[derive(Debug, WorldInit)]
struct ApiWorld {
    knative_wrapper: KnativeSpecWrapper,
    insta_label: String,
}

#[async_trait(?Send)]
impl World for ApiWorld {
    type Error = Infallible;

    async fn new() -> Result<Self, Self::Error> {
        Ok(ApiWorld {
            knative_wrapper: KnativeSpecWrapper(KnativeSpec::default()),
            insta_label: "label".to_owned(),
        })
    }
}

#[given(expr = "No set var to environment, use default ApiWorld")]
fn get_gvk_and_ksvc(_w: &mut ApiWorld) {}

#[given(expr = "Port and Image")]
fn port_and_image(w: &mut ApiWorld, step: &Step) {
    if let Some(table) = step.table.as_ref() {
        for row in table.rows.iter().skip(1) {
            // NOTE: skip header
            let port = &row[0];
            let image = &row[1];

            std::env::set_var("PORT", port);
            std::env::set_var("IMAGE", image);
            w.knative_wrapper = KnativeSpecWrapper(KnativeSpec::init_from_env().unwrap());
            w.insta_label = format!("{}-{}", port, image);
        }
    }
}

// !error from let wrapper = KnativeSpecWrapper(&w.knative_spec.clone());
// !use let wrapper = &w.knative_wrapper; and World{knative_wrapper: KnativeSpecWrapper,}

#[then(expr = "I get api")]
async fn show_api(w: &mut ApiWorld) {
    let wrapper = &w.knative_wrapper;
    insta_util::set_snapshot_suffix!("{}", w.insta_label);
    assert_debug_snapshot!(wrapper);

    let api = wrapper.get_api().await;
    assert_debug_snapshot!(api);
}

mod insta_util {
    macro_rules! set_snapshot_suffix {
        ($($expr:expr),*) => {{
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_suffix(format!($($expr,)*));
            settings.bind_to_thread();
        }}
    }
    pub(crate) use set_snapshot_suffix;
}
// #[macro_use]

#[tokio::main]
async fn main() {
    ApiWorld::run("tests/features/get_api.feature").await;
}
