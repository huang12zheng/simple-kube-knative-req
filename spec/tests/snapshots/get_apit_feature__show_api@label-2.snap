---
source: spec/tests/get_apit_feature.rs
expression: api
---
Api {
    request: Request {
        url_path: "/apis/serving.knative.dev/v1/namespaces/default/services",
    },
    client: "...",
    namespace: Some(
        "default",
    ),
}
